Table sorting across multiple pages, sort snapshots by recency

edlerd opened the pull request:
## Done

- fix instance and snapshot table sorting, when more than one page is available
- order snapshot table by recency

Fixes WD-3052

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @lorumic or @edlerd for access.
    - With a local copy of this branch, run as described in the [Readme](https://github.com/canonical/lxd-ui#setting-up-for-development).
2. Perform the following QA steps:
    - sort instances and snapshots when more than one page of items exists

webteam-app commented:
Demo starting at https://lxd-ui-289.demos.haus

al-bakalova commented:
@edlerd A few observations:

- Sorting the instance table by name works as expected
- Sorting by description seems unnecessary and is not working right now

- When sorting the different columns, the table "jumps" a couple of pixels. The issue is solved for Snapshot but is persistent in Instances.

- When sorting the snapshot table by name, we still sort only the list in the current page rather than the exhaustive list

edlerd commented:
> * Sorting by description seems unnecessary and is not working right now

Fixed sorting by description, there was missing a sort criterion for it.

> * When sorting the different columns, the table "jumps" a couple of pixels. The issue is solved for Snapshot but is persistent in Instances.

Fixed the height of the headings in the instance list.

> * When sorting the snapshot table by name, we still sort only the list in the current page rather than the exhaustive list

I think this is working as expected. The problem you might see is, we are sorting a mix of letters and numbers. So snap-100 comes before snap-50.

lorumic commented on src/pages/instances/InstanceSnapshots.tsx:
```diff
@@ -287,6 +287,9 @@ const InstanceSnapshots: FC<Props> = ({ instance }) => {
               instance.snapshots?.map((snapshot) => snapshot.name) ?? []
             }
+            onUpdateSort={pagination.updateSort}
+            defaultSort="created_at"
+            defaultSortDirection="descending"
```
For small screens, this is currently working only when you open the Snapshots tab. Then, as soon as you alter the sorting of any column, there's no way to "restore" this default way of sorting the data. It will be either sorted by name - even though the "arrow" is placed next to "Date created", which is confusing - or, if you "reset" the sort status of this column (= no arrow visible anymore), it will be sorted by date created but ascending instead of descending:

![Peek 2023-04-14 10-50](https://user-images.githubusercontent.com/56583786/231995890-95601703-9655-407d-9f2f-943b5899ce5a.gif)

edlerd replied:
This is a little tricky to solve. Some options that come to mind:
* Sort the collapsed column by date created instead of name
* Change which columns are collapsed, and not allow sorting by the collapsed column
* Disallow sorting the collapsed column as it is

edlerd replied:
I applied the 1. Idea and think it is actually pretty neat.

edlerd commented:
> * The clickable space between the snapshots and the status (instance table)

I just found a way to avoid this. Should now be limited to the text "Status".

> * Ignoring the capitalisation (name column)

This was already working. I realize in snapshots it doesn't work nicely, because we have a mixture of text and numbers. And I think that is quite a frequent use case with automated snapshots. So I disabled sort by name, the default sort by creation date should be more useful anyways.

al-bakalova commented:
@edlerd Looks good to me, thanks for applying the changes. The instance table performs as expected.

However, there seem to be an issue with the Name column of the Snapshots table. I am somehow still able to click on it to remove the sorting from the other columns but then I never get an ascending or a descending arrow. The text never turns blue either. The mouse pointer stays at "type pointer" and I believe the column is not sorted either. Could you please check on your side?

edlerd commented:
> However, there seem to be an issue with the Name column of the Snapshots table. I am somehow still able to click on it to remove the sorting from the other columns but then I never get an ascending or a descending arrow. The text never turns blue either. The mouse pointer stays at "type pointer" and I believe the column is not sorted either. Could you please check on your side?

This is a bug in the `MainTable` in react component. I created a [PR](https://github.com/canonical/react-components/pull/913) to fix it there. We have to keep it as is in our project until that is merged.

al-bakalova commented:
@edlerd Got it. It makes sense! From my point of view, we can proceed with the code review. I will give it a Design +1. Please let me know when this problem is solved for us so that I can check it again :)

lorumic commented:
> * When sorting the different columns, the table "jumps" a couple of pixels. The issue is solved for Snapshot but is persistent in Instances.

I'm still experiencing this "jump" issue on the collapsed snapshot table:

![Peek 2023-04-17 13-03](https://user-images.githubusercontent.com/56583786/232466479-1f820dbe-cfa3-456c-827f-56ab942bb00d.gif)

edlerd commented:
> I'm still experiencing this "jump" issue on the collapsed snapshot table:

Good catch on this collapsed state! I added a fix for it.

lorumic commented on src/pages/instances/InstanceSnapshots.tsx:
```diff
@@ -178,15 +178,15 @@ const InstanceSnapshots: FC<Props> = ({ instance }) => {
       ],
       sortData: {
-        name: snapshot.name,
+        name: snapshot.name.toLowerCase(),
```
We don't need this anymore, right? The table cannot be sorted by name.

edlerd replied:
true, good catch.

lorumic commented on src/util/pagination.tsx:
```diff
@@ -17,18 +18,51 @@ export const paginationOptions = [
   setPageSize: (newSize: number) => void;
   totalPages: number;
-} => {
+  updateSort: (sort: string | null | undefined) => void;
```
I think this can be shortened to:

```suggestion
  updateSort: (sort?: string) => void;
```

lorumic replied:
Wouldn't it still be compatible? I tested it locally and it seemed to work.

lorumic replied:
True, with `sort?: string | null` it's already better anyway.

lorumic commented on src/util/pagination.tsx:
```diff
@@ -17,18 +18,51 @@ export const paginationOptions = [
+export const usePagination = (
+  data: MainTableRow[],
+  defaultSort?: string,
+  defaultSortDirection?: SortDirection
+): Pagination => {
   const [_pageSize, _setPageSize] = useState(paginationOptions[0].value);
 
+  const [sort, setSort] = useState<string | null | undefined>(defaultSort);
+  const [sortDirection, setSortDirection] = useState(
+    defaultSortDirection ?? "ascending"
+  );
```
We can have an initializer, and get rid of the `??` operator in the state:

```suggestion
  defaultSortDirection: SortDirection = "ascending"
): Pagination => {
  const [_pageSize, _setPageSize] = useState(paginationOptions[0].value);

  const [sort, setSort] = useState<string | null | undefined>(defaultSort);
  const [sortDirection, setSortDirection] = useState(defaultSortDirection);
```

edlerd replied:
ah right, good idea.

lorumic commented on src/util/pagination.tsx:
```diff
@@ -17,18 +18,51 @@ export const paginationOptions = [
+    });
+  }
+
+  const updateSort = (newSort: string | null | undefined) => {
```
See comment above, about the interface for this function:
```suggestion
  const updateSort = (newSort?: string) => {
```
